File: agent/sap_nexus_agent/action_result.py
from __future__ import annotations

from dataclasses import dataclass
from typing import Any


@dataclass(frozen=True)
class ActionResult:
    trace_id: str
    capability_id: str
    success: bool
    pr_number: str
    commit_status: str
    return_messages: list[dict[str, Any]]
    duration_ms: int
    error_type: str

    @classmethod
    def from_dict(cls, payload: dict[str, Any]) -> "ActionResult":
        return cls(
            trace_id=str(payload.get("traceId", "")),
            capability_id=str(payload.get("capabilityId", "")),
            success=bool(payload.get("success", False)),
            pr_number=str(payload.get("prNumber", "")),
            commit_status=str(payload.get("commitStatus", "none")),
            return_messages=[dict(message) for message in payload.get("returnMessages", [])],
            duration_ms=int(payload.get("durationMs", 0)),
            error_type=str(payload.get("errorType", "UNKNOWN")),
        )

File: services/gateway/core/src/main/java/com/sapnexus/gateway/result/ActionResult.java
package com.sapnexus.gateway.result;

import java.util.List;

public record ActionResult(
        String traceId,
        String capabilityId,
        boolean success,
        String prNumber,
        CommitStatus commitStatus,
        List<SapReturnMessage> returnMessages,
        long durationMs,
        ErrorType errorType
) {
    public static ActionResult fromExecutionResult(ExecutionResult result) {
        Object rawPrNumber = result.data().get("prNumber");
        String prNumber = rawPrNumber == null ? "" : String.valueOf(rawPrNumber);
        CommitStatus status = commitStatusFrom(result);
        return new ActionResult(
                result.traceId(),
                result.capabilityId(),
                result.success(),
                prNumber,
                status,
                result.returnMessages(),
                result.durationMs(),
                result.errorType()
        );
    }

    public static ActionResult success(
            String traceId,
            String capabilityId,
            String prNumber,
            List<SapReturnMessage> returnMessages,
            long durationMs
    ) {
        return new ActionResult(
                traceId,
                capabilityId,
                true,
                prNumber,
                CommitStatus.committed,
                returnMessages,
                durationMs,
                ErrorType.NONE
        );
    }

    public static ActionResult failure(
            String traceId,
            String capabilityId,
            ErrorType errorType,
            String message,
            long durationMs
    ) {
        return new ActionResult(
                traceId,
                capabilityId,
                false,
                "",
                CommitStatus.none,
                List.of(new SapReturnMessage("E", "", "", message, "")),
                durationMs,
                errorType
        );
    }

    private static CommitStatus commitStatusFrom(ExecutionResult result) {
        Object rawStatus = result.data().get("commitStatus");
        if (rawStatus == null) {
            return CommitStatus.none;
        }
        try {
            return CommitStatus.valueOf(String.valueOf(rawStatus));
        } catch (IllegalArgumentException exception) {
            return CommitStatus.none;
        }
    }
}
